File: android/src/main/java/com/stallion/utils/StallionBSPatch.java
package com.stallion.utils;

public class StallionBSPatch {
  
  static {
    System.loadLibrary("stallion-bspatch");
  }
  
  /**
   * Applies a bsdiff patch file to oldFile, writing the result to newFile.
   * 
   * @param oldFile Path to the old/base file
   * @param newFile Path where the patched file will be written
   * @param patchFile Path to the patch file
   * @return true if patch was applied successfully, false otherwise
   */
  public static boolean applyPatch(String oldFile, String newFile, String patchFile) {
    int result = nativeApplyPatch(oldFile, newFile, patchFile);
    if (result == 0) {
      return true;
    } else {
      android.util.Log.e("StallionBSPatch", "BSPatch failed with code: " + result + 
          " (old: " + oldFile + ", patch: " + patchFile + ")");
      return false;
    }
  }
  
  private static native int nativeApplyPatch(String oldFile, String newFile, String patchFile);
}

File: ios/main/StallionBSPatch.swift
//
//  StallionBSPatch.swift
//  react-native-stallion
//
//  Created by Thor963 on 04/11/25.
//

import Foundation

class StallionBSPatch {
    
    static func applyPatch(oldFile: String, newFile: String, patchFile: String) -> Bool {
        #if os(iOS)
        // Use embedded C implementation on iOS
        // Swift automatically bridges String to const char * for C functions
        let result = oldFile.withCString { oldPath in
            newFile.withCString { newPath in
                patchFile.withCString { patchPath in
                    bspatch_apply(oldPath, newPath, patchPath)
                }
            }
        }
        
        if result == 0 {
            return true
        } else {
            NSLog("BSPatch failed with code: %d (old: %@, patch: %@)", result, oldFile, patchFile)
            return false
        }
        #else
        NSLog("BSPatch not supported on this platform.")
        return false
        #endif
    }
}
